### PR TITLE
rocAL - External Source Reader

### PR DESCRIPTION
**Pull request by swetha097**
Has 3 Modes in rocAL whose sources varies accordingly :
• Mode 0 - File Names
• Mode 1 - Raw Buffer
• Mode 2 - Decoded Buffer

• Python Support Given for Mode 0 only.
• C++ support has the support for all 3 Modes
Note: This includes the changes from rocAL - performance PR

### CODE REVIEW

**Comment by kiritigowda** [comment]
@LakshmiKumar23 is this ready to be merged?

**Comment by kiritigowda** [comment]
@LakshmiKumar23 is this ready to be merged?

File: rocAL/rocAL/include/api/rocal_api_data_transfer.h [hunk 1]
```diff
@@ -48,5 +48,23 @@ extern "C"  RocalStatus   ROCAL_API_CALL rocalCopyToOutputTensor(RocalContext ro
...
+/// \param roi_width The roi width of the images
+/// \param roi_height The roi height of the images
+/// \param max_width The maximum width of the decoded images, larger or smaller will be resized to closest
+/// \param max_height The maximum height of the decoded images, larger or smaller will be resized to closest
+/// \param channels The number of channels for the image
+/// \param mode Determines the mode of the source passed from the user - file_names / uncompressed data / compressed data
+/// \param layout Determines the layout of the images - NCHW / NHWC
+/// \return Reference to the output image
+extern "C" RocalStatus ROCAL_API_CALL rocalExternalSourceFeedInput(RocalContext rali_context,std::vector<std::string> input_images_names,
+                                                                std::vector<int> labels, std::vector<unsigned char *> input_buffer,
```
Comment (rrawther) [inline rocAL/rocAL/include/api/rocal_api_data_transfer.h:65]: why are the roi_width and roi_height is passed through this api?

File: rocAL/rocAL/include/loaders/image/image_loader.h [hunk 2]
```diff
@@ -83,6 +87,8 @@ class ImageLoader : public LoaderModule {
     size_t _image_counter = 0;//!< How many images have been loaded already
     size_t _remaining_image_count;//!< How many images are there yet to be loaded
     bool _decoder_keep_original = false;
+    bool _external_source_reader = false; //!< Set to true if external source reader
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/loaders/image/image_loader.h:90]: We are setting this to false in the master graph also. Why are we setting it separately only for the image_loader?

File: rocAL/rocAL/include/pipeline/commons.h [hunk 3]
```diff
@@ -95,7 +95,6 @@ struct Timing
     long long unsigned label_load_time= 0;
     long long unsigned bb_load_time= 0;
     long long unsigned mask_load_time = 0;
-    long long unsigned shuffle_time = 0;
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/pipeline/commons.h:98]: This has already been removed by Rajy in another PR. Not sure why it's seen in your branch. Please sync to TOT

File: rocAL/rocAL/include/pipeline/context.h [hunk 4]
```diff
@@ -35,7 +35,6 @@ struct Context
     {
         LOG("Processing on " + STR(((affinity == RocalAffinity::CPU)?" CPU": " GPU")))
         master_graph = std::make_shared<MasterGraph>(batch_size, affinity, gpu_id, prefetch_queue_depth, output_tensor_type);
-        _internal_batch_size = master_graph->internal_batch_size();
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/pipeline/context.h:38]: same as above

File: rocAL/rocAL/include/readers/image/external_source_reader.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    std::string id() override { return _last_id; }
+
+    //! Return batch_size() for count_items unless end_of_sequence has been signalled
+    unsigned count_items() override;
+
+    ~ExternalSourceReader() override;
+
+    int close() override;
+    // get_shuffle_time() not applicable for external_source reader
+    unsigned long long get_shuffle_time() { return 0; }
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/readers/image/external_source_reader.h:65]: Why do need get_shuffle_time? We got rid of it

File: rocAL/rocAL/include/readers/image/external_source_reader.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    bool _shuffle;
+    int _read_counter = 0;
+    volatile bool _end_of_sequence;
+    //!< _file_count_all_shards total_number of files in to figure out the max_batch_size (usually needed for distributed training).
+    void push_file_name(const std::string& image_name);
+    bool pop_file_name(std::string& file_name);
+    void push_file_data(std::tuple<unsigned char*, size_t, int, int, int>& image);
+    bool pop_file_data(std::tuple<unsigned char*, size_t, int, int, int>& image);
+    size_t  _file_count_all_shards;
+    void incremenet_read_ptr();
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/readers/image/external_source_reader.h:115]: spelling check for "increment", below also

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]
@swetha097 This PR seems to be missing changes already merged from other PRs. Specifically the  shuffle time changes and the user batch size changes. Please sync with TOT once and that should further reduce the number of file chnages

**Comment by shobana-mcw** [comment]
@LakshmiKumar23 As discussed in weekly call we will merge with TOT and inform you since some classifications changes are also present here. 
Please commence the review after we sync up with TOT.

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 7]
```diff
@@ -104,8 +103,14 @@ class MasterGraph
     bool is_video_loader() {return _is_video_loader; }
     bool is_sequence_reader_output() {return _is_sequence_reader_output; }
     void set_sequence_reader_output() { _is_sequence_reader_output = true; }
-    void set_sequence_batch_size(size_t sequence_length) { _sequence_batch_size = _user_batch_size * sequence_length; }
-    void set_sequence_batch_ratio() { _sequence_batch_ratio = _sequence_batch_size / _internal_batch_size; }
+    void set_sequence_batch_size(size_t sequence_length) {
```
Comment (rrawther) [inline rocAL/rocAL/include/pipeline/master_graph.h:106]: please remove empty lines. Why is this function changed for external source

File: rocAL/rocAL/include/readers/image/external_source_reader.h (hunk 6 above)
Comment (swetha097) [reply rocAL/rocAL/include/readers/image/external_source_reader.h]: Resolved

**Comment by swetha097** [comment]
@rrawther @LakshmiKumar23 
Closing this PR as it shows changes from other merged PR's even after syncing the branch with the latest TOT.
Will issue a new one.

**Pull request closed without merging**